web/src/containers/SaveButton.js

```diff
@@ -0,0 +1,126 @@
+import { Alert, Button, Spinner } from '@cmsgov/design-system-core';
+import PropTypes from 'prop-types';
+import React, { Fragment, PureComponent } from 'react';
+import { connect } from 'react-redux';
+
+import { saveApd } from '../actions/apd';
+import { getIsDirty } from '../reducers/dirty';
+import { getSaveApdError } from '../reducers/errors';
+import { getSaveApdWorking } from '../reducers/working';
+
+const getButtonContent = (dirty, working) => {
+  if (!dirty) {
+    return 'Saved';
+  }
+
+  if (working) {
+    return (
+      <Fragment>
+        <Spinner /> Saving
+      </Fragment>
+    );
+  }
+
+  return 'Save';
+};
+
+class SaveButton extends PureComponent {
+  state = {
+    hasFetched: false,
+    success: false
+  };
+
+  static getDerivedStateFromProps({ error, working }, { hasFetched }) {
+    // Success has to be derived.  It can't be stored in the app state because
+    // if it was, then the next time this form was loaded, it would show the
+    // success state even though it wouldn't be accurate anymore.
+    if (hasFetched) {
+      const success = !working && !error;
+      return { success };
+    }
+    return null;
+  }
+
+  setSuccessTimer = (() => {
+    let alertTimer = null;
+    return () => {
+      if (alertTimer) {
+        // If there's already a save timer, clear it or else it'll close the
+        // success dialog when it runs out rather than five seconds from now
+        clearTimeout(alertTimer);
+      }
+      alertTimer = setTimeout(this.clearFetched, 5000);
+    };
+  })();
+
+  clearFetched = () => {
+    this.setState({ hasFetched: false });
+  };
+
+  save = () => {
+    const { saveApd: action } = this.props;
+    this.setState({ hasFetched: true });
+    action();
+  };
+
+  render() {
+    const { dirty, error, working } = this.props;
+    const { hasFetched, success } = this.state;
+
+    let alert = null;
+    if (hasFetched) {
+      if (error) {
+        alert = (
+          <Alert variation="error" className="ds-u-margin-bottom--2">
+            {error}
+          </Alert>
+        );
+      }
+      if (success) {
+        alert = (
+          <Alert variation="success" className="ds-u-margin-bottom--2">
+            Save successful!
+          </Alert>
+        );
+        this.setSuccessTimer();
+      }
+    }
+
+    const buttonVariant = dirty ? 'primary' : 'success';
+
+    return (
+      <div className="ds-l-row save-button--container">
+        <div className="ds-l-col--3">
+          <div className="ds-u-margin-bottom--3">
+            <div className="save-button--alert">{alert}</div>
+            <Button variation={buttonVariant} onClick={this.save}>
```

Comment:
So my thinking was if I put this in a row, then I could have an offset column pushed to the right and that'd align things, but it turned out that didn't work like I thought.  Once the button's parent element was `fixed`, it no longer obeyed the alignment of the column.

If I'm completely honest, I don't know which of these various `<div>`s are actually necessary at this point.  The `ds-l-col--3` was meant to constrain the alerts so they didn't just grow forever, but that didn't work so I set the width directly - and now I'm not sure if this column class is doing anything.